diff --git a/alloc.c b/alloc.c
index a53e12c0..6e7be764 100644
--- a/alloc.c
+++ b/alloc.c
@@ -19,7 +19,7 @@
 #include "private/gc_priv.h"
 
 #include <stdio.h>
-#if !defined(MACOS) && !defined(MSWINCE)
+#if !defined(MACOS) && !defined(MSWINCE) && !defined(WASI)
 # include <signal.h>
 # if !defined(GC_NO_TYPES) && !defined(SN_TARGET_PSP2) \
      && !defined(__CC_ARM)
diff --git a/include/private/gc_priv.h b/include/private/gc_priv.h
index 2ffd4dd0..ef5a7007 100644
--- a/include/private/gc_priv.h
+++ b/include/private/gc_priv.h
@@ -838,7 +838,9 @@ EXTERN_C_END
 # endif
 #endif /* DARWIN */
 
-#include <setjmp.h>
+#if !defined(WASI)
+# include <setjmp.h>
+#endif
 
 #if __STDC_VERSION__ >= 201112L
 # include <assert.h> /* for static_assert */
diff --git a/include/private/gcconfig.h b/include/private/gcconfig.h
index 2de01afe..cf25e78c 100644
--- a/include/private/gcconfig.h
+++ b/include/private/gcconfig.h
@@ -703,6 +703,11 @@ EXTERN_C_BEGIN
 #   define I386
 #   define mach_type_known
 # endif
+# if defined(__wasi__)
+#   define WASI
+#   define I386
+#   define mach_type_known
+#endif
 
 /* Feel free to add more clauses here */
 
@@ -1425,6 +1430,21 @@ EXTERN_C_BEGIN
 #     define USE_MMAP_ANON      /* avoid /dev/zero, not supported */
 #     define STACK_GROWS_DOWN
 #   endif
+#   ifdef WASI
+      extern unsigned char __global_base;
+      extern unsigned char __heap_base;
+#     define OS_TYPE "WASI"
+#     define DATASTART ((ptr_t)&__global_base)
+#     define DATAEND ((ptr_t)&__heap_base)
+#     define STACKBOTTOM ((ptr_t)&__global_base)
+#     define GETPAGESIZE() 65536
+#     define USE_MMAP_ANON      /* avoid /dev/zero, not supported */
+#     define GC_NO_SIGSETJMP 1
+#     define NO_CLOCK 1
+#     ifndef HBLKSIZE
+#       define HBLKSIZE 4096
+#     endif
+#   endif
 #   if defined(__QNX__)
 #     define OS_TYPE "QNX"
 #     define SA_RESTART 0
diff --git a/mark.c b/mark.c
index b33bb171..25045f5d 100644
--- a/mark.c
+++ b/mark.c
@@ -1586,6 +1586,7 @@ GC_API void GC_CALL GC_push_all_eager(void *bottom, void *top)
         REGISTER word q;
 
         LOAD_WORD_OR_CONTINUE(q, p);
+        GC_printf("%p %p\n", (void*)p, (void*)q);
         GC_PUSH_ONE_STACK(q, p);
       }
 #   undef GC_greatest_plausible_heap_addr
diff --git a/mark_rts.c b/mark_rts.c
index a0d67641..af5f2665 100644
--- a/mark_rts.c
+++ b/mark_rts.c
@@ -734,8 +734,10 @@ STATIC void GC_push_all_stack_partially_eager(ptr_t bottom, ptr_t top,
     GC_ASSERT((word)bottom <= (word)cold_gc_frame
               && (word)cold_gc_frame <= (word)top);
 #   ifdef STACK_GROWS_DOWN
-        GC_push_all(cold_gc_frame - sizeof(ptr_t), top);
+        GC_push_all_eager(cold_gc_frame - sizeof(ptr_t), top);
+        GC_printf("***push_all from %p to %p\n", cold_gc_frame - sizeof(ptr_t), top);
         GC_push_all_eager(bottom, cold_gc_frame);
+        GC_printf("***push_all_eager from %p to %p\n", bottom, cold_gc_frame);
 #   else /* STACK_GROWS_UP */
         GC_push_all(bottom, cold_gc_frame + sizeof(ptr_t));
         GC_push_all_eager(cold_gc_frame, top);
diff --git a/misc.c b/misc.c
index 64eeb7e4..406d5b93 100644
--- a/misc.c
+++ b/misc.c
@@ -20,7 +20,7 @@
 #include <limits.h>
 #include <stdarg.h>
 
-#ifndef MSWINCE
+#if !defined(MSWINCE) && !defined(WASI)
 # include <signal.h>
 #endif
 
@@ -823,7 +823,7 @@ GC_API int GC_CALL GC_is_init_called(void)
 #if !defined(OS2) && !defined(MACOS) && !defined(GC_ANDROID_LOG) \
     && !defined(NN_PLATFORM_CTR) && !defined(NINTENDO_SWITCH) \
     && (!defined(MSWIN32) || defined(CONSOLE_LOG)) && !defined(MSWINCE)
-  STATIC int GC_stdout = GC_DEFAULT_STDOUT_FD;
+  STATIC int GC_stdout = GC_DEFAULT_STDERR_FD;
   STATIC int GC_stderr = GC_DEFAULT_STDERR_FD;
   STATIC int GC_log = GC_DEFAULT_STDERR_FD;
 
diff --git a/os_dep.c b/os_dep.c
index e116ad02..4bd9031c 100644
--- a/os_dep.c
+++ b/os_dep.c
@@ -27,14 +27,14 @@
 #endif
 
 #include <stdio.h>
-#if defined(MSWINCE) || defined(SN_TARGET_PS3)
+#if defined(MSWINCE) || defined(SN_TARGET_PS3) || defined(WASI)
 # define SIGSEGV 0 /* value is irrelevant */
 #else
 # include <signal.h>
 #endif
 
 #if defined(UNIX_LIKE) || defined(CYGWIN32) || defined(NACL) \
-    || defined(SYMBIAN)
+    || defined(SYMBIAN) || defined(WASI)
 # include <fcntl.h>
 #endif
 
@@ -2668,7 +2668,7 @@ static void block_unmap_inner(ptr_t start_addr, size_t len)
             if (madvise(start_addr, len, MADV_DONTNEED) == -1)
               ABORT_ON_REMAP_FAIL("unmap: madvise", start_addr, len);
 #         endif
-#       elif defined(EMSCRIPTEN)
+#       elif defined(EMSCRIPTEN) || defined(WASI)
           /* Nothing to do, mmap(PROT_NONE) is not supported and        */
           /* mprotect() is just a no-op.                                */
 #       else
