diff --git a/src/cre2.cpp b/src/cre2.cpp
index 5a63b93..12bc2ed 100644
--- a/src/cre2.cpp
+++ b/src/cre2.cpp
@@ -462,7 +462,6 @@ DEFINE_MATCH_REX_FUN2(cre2_find_and_consume_re,FindAndConsumeN)
 int
 cre2_replace (const char * pattern, cre2_string_t * text_and_target, cre2_string_t * rewrite)
 {
-  try {
     std::string		S(text_and_target->data, text_and_target->length);
     re2::StringPiece	R(rewrite->data, rewrite->length);
     char *		buffer; /* this exists to make GCC shut up about const */
@@ -477,12 +476,6 @@ cre2_replace (const char * pattern, cre2_string_t * text_and_target, cre2_string
     } else
       return -1;
     return int(retval);
-  } catch(const std::exception &e) {
-    // e.what();
-    return -1;
-  } catch(...) {
-    return -1;
-  }
 }
 int
 cre2_replace_re (cre2_regexp_t * rex, cre2_string_t * text_and_target, cre2_string_t * rewrite)
